liferay-data-engine/liferay-portal
# LPS-165782 Keyboard DND

Motivation
=======
Allowing drag and drop with keyboard, as reported in https://issues.liferay.com/browse/LPS-165782

Proposed Solution
========
Basically I have mimicked the behavior that we have implemented in Page Editor, as it is working fine. As a general idea, there is a new React Context called `KeyboardDNDContext` that maintains the whole logic for DND, and the rest of the code is connected using a couple of hooks to begin the drag process and handle hover state.

Once the drag process has been started, a `keydown` event listener is attached to the document, preventing all keyboard navigation until:
- The `KeyboardDNDContext` is unmounted (user navigates out of the page).
- User presses `Enter` key to drop the item in the desired position.
- User presses `Escape` key to abort the drag process.

The KeyboardDNDContext has two elements with the dragging information:
- `sourceItem`: contains all needed information about the item that is being dragged.
- `currentTarget`: information about where the item is going to be dropped.

The key aspect about `currentTarget` is that always contains a valid drop position, so when a user presses `Enter` key, we know that we can drop the item in that place. To ensure that this always happen, we have created a `getNextTarget` function inside `KeyboardDNDContext` that:

1. Navigates to the next available position depending on the pressed key.
2. Checks if that position is valid and:
   - If `true`, it returns the position.
   - If `false`, (recursion) it continues navigating in the same direction until
     - A valid position is found.
     - The navigation reaches page end (returning the starting position).

The hardest part of the implementation has been creating a unique ID for every item on the form. Since we do not have IDs for every elements (only for Fields, but not columns, rows, pages...), we have added a new `itemPath` property that is constructed through all UI components, representing the index (position) of each element in the `pages` property of `FormState`, for example, for `itemPath = [0, 5, 1, 2, 3, 2]` that means:

```
First page
 > Sixth row
 > Second column
 > Third field
 > Fourth row // This means that this field has nested fields
 > Second column
```

With this information we have been able to navigate through the form structure.

There is a pending task that should be checked, and it's about keyboard focus management. In this PR I that tried not manipulating the keyboard focus too much, since I do not want to cause any conflict with the behavior of the application. However, it would be nice to analyze and check in the future where the keyboard focus should be moved after adding/moving an item with keyboard.

Also, the active/hover feedback should be checked, as sometimes there are hovered/selected/keyboard focused elements at the same time, which might be a bit confusing for users (they have the same visual feedback).  

Steps to Verify
========
1. Deploy all modules in this PR
1. Focus some item inside the sidebar (using tab).
1. Press `Space` key (drag and drop starts, see that a drop feedback will appear at the end of the page).
1. Move the target with `up/down` arrow keys.
1. Press enter to drop the item.

The same process can be followed to move items, but in this case a hidden drag handler in the topper can be focused with keyboard.

References to existing code
========
https://github.com/liferay/liferay-portal/blob/master/modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/KeyboardMovementManager.js

Screenshots
========
![Peek 2022-11-03 17-38](https://user-images.githubusercontent.com/800645/199781079-3524f03a-4bce-4b7c-96e2-e42399fc2a50.gif)

Checklist
========

CODE:

COMMITS:

PR TITLE & DESCRIPTION:
